src/windows/list.cpp

```diff
@@ -0,0 +1,687 @@
+      break;
+    }
+
+    if (!GetDeviceSize(hPhysical, device)) {
+      errorCode = GetLastError();
+      device->error = "Couldn't get disk geometry: Error " +
+        std::to_string(errorCode);
+      result = false;
+      break;
+    }
+
+    if (!GetAdapterInfo(hPhysical, device)) {
+      errorCode = GetLastError();
+      device->error = "Couldn't get device adapter descriptor: Error " +
+        std::to_string(errorCode);
+      result = false;
+      break;
+    }
+
+    // NOTE: No need to fail over this one,
+    // as we can safely default to a 512B block size
+    if (!GetDeviceBlockSize(hPhysical, device)) {
+      errorCode = GetLastError();
+      printf("[INFO] Couldn't get block size: Error %u\n", errorCode);
+    }
+
+    BOOL isWritable = DeviceIoControl(
+      hPhysical, IOCTL_DISK_IS_WRITABLE, NULL, 0,
+      NULL, 0, &size, NULL);
+
+    device->isReadOnly = !isWritable;
+  }
+
+  if (hDevice != INVALID_HANDLE_VALUE) {
+    CloseHandle(hDevice);
+  }
+
+  if (hPhysical != INVALID_HANDLE_VALUE) {
+    CloseHandle(hPhysical);
+  }
+
+  free(deviceDetailData);
+
+  return result;
+}
+
+std::vector<DeviceDescriptor> ListStorageDevices() {
+  HDEVINFO hDeviceInfo = NULL;
+  SP_DEVINFO_DATA deviceInfoData;
+  std::vector<DeviceDescriptor> deviceList;
+
+  DWORD i;
+  char *enumeratorName;
+  DeviceDescriptor device;
+
+  hDeviceInfo = SetupDiGetClassDevsA(
+    &GUID_DEVICE_INTERFACE_DISK, NULL, NULL,
+    DIGCF_PRESENT | DIGCF_DEVICEINTERFACE);
+
+  if (hDeviceInfo == INVALID_HANDLE_VALUE) {
+    printf("[ERROR] Invalid DeviceInfo handle\n");
+    return deviceList;
+  }
+
+  deviceInfoData.cbSize = sizeof(SP_DEVINFO_DATA);
+
+  for (i = 0; SetupDiEnumDeviceInfo(hDeviceInfo, i, &deviceInfoData); i++) {
+    enumeratorName = GetEnumeratorName(hDeviceInfo, deviceInfoData);
+
+    // If it failed to get the SPDRP_ENUMERATOR_NAME, skip it
+    if (enumeratorName == NULL) {
+      continue;
+    }
+
+    device = DeviceDescriptor();
+
+    device.enumerator = std::string(enumeratorName);
+    device.description = GetFriendlyName(hDeviceInfo, deviceInfoData);
+    device.isRemovable = IsRemovableDevice(hDeviceInfo, deviceInfoData);
+    device.isVirtual = IsVirtualHardDrive(hDeviceInfo, deviceInfoData);
+    device.isSCSI = IsSCSIDevice(enumeratorName);
+    device.isUSB = IsUSBDevice(enumeratorName);
+    device.isCard = device.enumerator == "SD";
+    // device.isSystem = IsSystemDevice(hDeviceInfo, deviceInfoData);
+    device.isSystem = device.isSCSI && !device.isRemovable;
+    device.isUAS = device.isSCSI && device.isRemovable &&
+      !device.isVirtual && !device.isCard;
+
+    if (GetDetailData(&device, hDeviceInfo, deviceInfoData)) {
+      device.isCard = device.busType == "SD" || device.busType == "MMC";
+      device.isUAS = device.enumerator == "SCSI" && device.busType == "USB";
+      device.isVirtual = device.isVirtual ||
+        device.busType == "VIRTUAL" ||
+        device.busType == "FILEBACKEDVIRTUAL";
```

Comment:
I don't mean to sound pessimistic, but it feels like a lot of these `isXXXX` properties are based on guesswork and heuristics, so perhaps they're not as reliable as we'd want them to be?  (unless of course they get documented as possibly not being reliable)

Reply:
How is this a heuristic? If the bus type is virtual, or the hardware ID is that of a known virtual device, it is virtual. That's no guesswork, only that the previous check works without GetDetailData(). And since GetDetailData() might error for various reasons, it's of interest to check what we can beforehand, and populate the properties, to have more info for debugging.

Reply:
I wasn't specifically referring to the `.isVirtual` method, more about the `.isXXXX` properties in general, e.g.
```
device.isUAS = device.isSCSI && device.isRemovable && !device.isVirtual && !device.isCard;
```
and
```
device.isUAS = device.enumerator == "SCSI" && device.busType == "USB";
```
and the fact that `IsSCSIDevice` and `IsUSBDevice` themselves only work by attempting to match a `enumeratorName` against a (possibly incomplete) list of hardcoded strings.
But like I said, perhaps I'm being too pessimistic ;-)